src/tqec/compile/substitute.py

```diff
@@ -0,0 +1,203 @@
+    else:
+        return _substitute_in_space_with_usual_cubes(key, block1, block2)
+
+
+DEFAULT_SUBSTITUTION_RULES: defaultdict[SubstitutionKey, SubstitutionRule] = (
+    defaultdict(lambda: default_substitution_rule)
+)
+
+
+def _substitute_in_time_direction(
+    key: SubstitutionKey,
+    block1: CompiledBlock,
+    block2: CompiledBlock,
+) -> tuple[CompiledBlock, CompiledBlock]:
+    assert key.pipe_type.direction == Direction3D.Z, "Pipe direction must be Z."
+    # Substitute the final layer of the bottom block with the bulk layer
+    bottom_block = block1.with_updated_layer(
+        plaquettes_to_update=deepcopy(block1.layers[1].collection),
+        layer_to_update=block1.num_layers - 1,
+    )
+    # Substitute the first layer of the top block with the bulk layer
+    top_block = block2.with_updated_layer(
+        plaquettes_to_update=deepcopy(block2.layers[1].collection),
+        layer_to_update=0,
+    )
+    # If the pipe has a hadamard, apply the hadamard transformation
+    # The hadamard transformation is applied to the end of the bottom block
+    if key.pipe_type.has_hadamard:
+        raise TQECException("Pipe with hadamard is not supported yet.")
+    return bottom_block, top_block
+
+
+def _substitute_in_space_with_usual_cubes(
+    key: SubstitutionKey,
+    block1: CompiledBlock,
+    block2: CompiledBlock,
+) -> tuple[CompiledBlock, CompiledBlock]:
+    pipe_type = key.pipe_type
+    assert (
+        not key.spec1.is_spatial_junction and not key.spec2.is_spatial_junction
+    ), "Both cubes must not be spatial junction."
+    assert pipe_type.direction != Direction3D.Z, "Pipe direction must be X or Y."
+    if pipe_type.has_hadamard:
+        raise TQECException("Pipe with hadamard is not supported yet.")
+    substitute_side1 = (
+        PlaquetteSide.RIGHT
+        if pipe_type.direction == Direction3D.X
+        else PlaquetteSide.DOWN
+    )
+    substitute_side2 = substitute_side1.opposite()
+    # Get reset/mesurement basis
+    basis = pipe_type.temporal_basis()
+    block1 = _substitute_on_the_border(block1, substitute_side1, basis)
+    block2 = _substitute_on_the_border(block2, substitute_side2, basis)
+    return block1, block2
+
+
+def _substitute_on_the_border(
+    block: CompiledBlock,
+    plaquette_side: PlaquetteSide,
+    basis: Literal["X", "Z"],
+) -> CompiledBlock:
+    if plaquette_side == PlaquetteSide.LEFT:
+        substitution = {1: 6, 8: 9, 7: 10}
+    elif plaquette_side == PlaquetteSide.UP:
+        substitution = {2: 12, 5: 10, 6: 9}
+    elif plaquette_side == PlaquetteSide.DOWN:
+        substitution = {3: 7, 14: 10, 13: 9}
+    else:  # plaquette_side == PlaquetteSide.RIGHT:
+        substitution = {4: 13, 11: 9, 12: 10}
+    # 1. Fill in/Substitue the plaquettes on the border
+    for i in range(block.num_layers):
+        block = block.with_updated_layer(
+            plaquettes_to_update={
+                dst: deepcopy(block.layers[i][src]) for dst, src in substitution.items()
```

Comment:
We need it here because later on we will add reset/measurement on data qubits at certain layers. If we do not deep copy it, we will end up adding unwanted operations to some layers. I tested after removing it and can be sure about this.